### DISCUSSION TITLE
Direct attached with a USB to Serial Adapter

### DISCUSSION THREAD

**Discussion by Buddeli** [adorobis/hacomfoairmqtt#60] [General]
Hello everyone

I just tried to connect over a USB to serial adapter with the docker container option. Sadly I just reserved the following lines after crating the container:

hassioaddon-comfoair350  | don't create serial device over ehternet. enable it with SOCAT=True
hassioaddon-comfoair350  | *****************************
hassioaddon-comfoair350  | * CA350 MQTT Home Assistant *
hassioaddon-comfoair350  | *****************************
hassioaddon-comfoair350  | 
hassioaddon-comfoair350  | 31-07-2024 21:13:46 WARNING: Opening serial port exception:
hassioaddon-comfoair350  | 31-07-2024 21:13:46 WARNING: (<class 'serial.serialutil.SerialException'>, SerialException(2, 'could not open port "/dev/ttyUSB0": [Errno 2] No such file or directory: \'"/dev/ttyUSB0"\''), <traceback object at 0x78b168257d00>)

Here the configuration in the docker-compose.yml
version: '2'

services:    
  hassioaddon-comfoair350:
    image: k42sde/hassioaddon-comfoair350:latest
    container_name: hassioaddon-comfoair350
    environment:
      - SOCAT=False
      - COMFOAIR_IP=192.168.1.50
      - COMFOAIR_PORT=502
      - SERIAL_PORT="dev/ttyUSB0"
      - RS485_PROTOCOL=False
      - REFRESH_INTERVAL=10
      - ENABLE_PC_MODE=False
      - DEBUG=False
      - MQTT_SERVER=192.168.178.43
      - MQTT_PORT=1883
      - MQTT_KEEPALIVE=45
      - MQTT_USER=comfoair
      - MQTT_PASSWORD=XYZ
      - HA_ENABLE_AUTO_DISCOVERY_SENSORS=True
      - HA_ENABLE_AUTO_DISCOVERY_CLIMATE=True
      - HA_AUTO_DISCOVERY_DEVICE_ID=ca350
      - HA_AUTO_DISCOVERY_DEVICE_NAME=CA350
      - HA_AUTO_DISCOVERY_DEVICE_MANUFACTURER=Zehnder
      - HA_AUTO_DISCOVERY_DEVICE_MODEL=Comfoair 350
    restart: always

Do you have an idea why I got the error? Many thanks for answering.

**Comment by adorobis** [comment]
The error clearly says it cannot open the usb port. I am not using this docker deployment method but based on my experience you should pass the usb device from host to the container. Here is an example for my other use case (zigbee2mqtt):
```
    devices:
      - /dev/serial/by-id/usb-dresden_elektronik_ingenieurtechnik_GmbH_ConBee_II_DE2400281-if00:/dev/ttyACM0
```

so if your `dev/ttyUSB0` is a correct port it would look like that:
```
    devices:
      - /dev/ttyUSB0:/dev/ttyUSB0
```
Let me know if this works so I can update the wiki here.

**Reply by Buddeli** [reply]
I have now set up the Docker environment in Proxmox on a VM (previously it was an LXC). Now I can make sure that I can pass through the USB port without mistyping. Furthermore, I was able to ensure with the esptool (esptool flash_id) that the interface is seen in the VM:
_esptool.py v2.8
Found 1 serial ports
Serial port /dev/ttyUSB0)_

Unfortunately I still get the same message that the port cannot be opened. Would it be necessary to adapt the yml-file and point to the USB device?

**Reply by adorobis** [reply]
There is no `devices:` section in your docker compose yaml enrty. You need to define the serial device in its configuration like I've mentioned above. Moreover I don't see any reason why you would need to create a VM in which you'd run a docker. You can run docker right on the host and pass the USB device to the container.

**Reply by jschanz** [reply]
@Buddeli  .. Please add 

```
devices:
  /dev/ttyUSB0:/dev/ttyUSB0
```

to your docker-compose.yml.
As the docker image is actually based on Ubuntu, you can also enter the docker container and check if the device is visible under /dev/ttyUSB0 after adding the lines above.

`docker exec -it <container-name-or-id> /bin/bash`

Maybe the documentation is incomplete at this point, because I only use the remote serial option with socat.

**Reply by Buddeli** [reply]
Thanks for the hint. I have already tried the adjustment in the yml-file, but without success. I still get the same message. Do I need to make any further adjustments? I have also tried the option with "Privilege = true", also without success. Here is the customized yml file:

_services:    
  hassioaddon-comfoair350:
    image: k42sde/hassioaddon-comfoair350:latest
    container_name: hassioaddon-comfoair350
    environment:
      - SOCAT=False
      - COMFOAIR_IP=192.168.1.50
      - COMFOAIR_PORT=502
      - SERIAL_PORT="/dev/ttyUSB0"
      - RS485_PROTOCOL=False
      - REFRESH_INTERVAL=10
      - ENABLE_PC_MODE=False
      - DEBUG=False
      - MQTT_SERVER=192.168.178.43
      - MQTT_PORT=1883
      - MQTT_KEEPALIVE=45
      - MQTT_USER=comfoair
      - MQTT_PASSWORD=I don't tell you :)
      - HA_ENABLE_AUTO_DISCOVERY_SENSORS=True
      - HA_ENABLE_AUTO_DISCOVERY_CLIMATE=True
      - HA_AUTO_DISCOVERY_DEVICE_ID=ca350
      - HA_AUTO_DISCOVERY_DEVICE_NAME=CA350
      - HA_AUTO_DISCOVERY_DEVICE_MANUFACTURER=Zehnder
      - HA_AUTO_DISCOVERY_DEVICE_MODEL=Comfoair 350
    restart: always
    privileged: true
    devices:
      - /dev/ttyUSB0:/dev/ttyUSB0_

if I enter the command `docker exec -it <container-name-or-id> /bin/bash`, I receive the following answer: 
`Error response from daemon: Container 43b239c3a364dc7c3245d627639906be690862539c61159f0ffd3f952c485fb5 is restarting, wait until the container is running`

But in the portainer.io I see that the container is up and running.

**Reply by adorobis** [reply]
Check the logs of the container in portainer. Most likely it is restarting due to some errors which you'll see there.

**Reply by Buddeli** [reply]
Thank you for the quick response. In the log I can see the following comment:
```
_* CA350 MQTT Home Assistant *
*****************************
04-08-2024 19:39:20 WARNING: Opening serial port exception:
04-08-2024 19:39:20 WARNING: (<class 'serial.serialutil.SerialException'>, SerialException(2, 'could not open port "/dev/ttyUSB0": [Errno 2] No such file or directory: \'"/dev/ttyUSB0"\''), <traceback object at 0x7f7e8cc6bcc0>)
don't create serial device over ehternet. enable it with SOCAT=True_
```

**Reply by adorobis** [reply]
Well, the message is clear. No such device in your container. Are you sure this is the correct device name in your docker host (I suppose you still use the VM for it)? Check if that device indeed exists on host.

**Reply by Buddeli** [reply]
When I enter `ls -la /dev/serial/by-id/` I can see the device. 
```
buddy@dockerversuch:~$ ls -la /dev/serial/by-id/
total 0
drwxr-xr-x 2 root root 60 Aug  3 18:08 .
drwxr-xr-x 4 root root 80 Aug  3 18:08 ..
lrwxrwxrwx 1 root root 13 Aug  3 18:08 usb-Silicon_Labs_CP2102_USB_to_UART_Bridge_Controller_0001-if00-port0 -> ../../ttyUSB0
``` 

Even with the esptool I can find the USB Device:

```
buddy@dockerversuch:~$ esptool flash_id
esptool.py v2.8
Found 1 serial ports
Serial port /dev/ttyUSB0
Connecting...
```

**Comment by jschanz** [comment]
@Buddeli ... Could you please provide the content of the file "/opt/hacomfoairmqtt/config.ini". 

The SOCAT=True message is triggered by start.sh:

```
# Start the first process
if [ "$SOCAT" == "True" ]; then
    echo "create serial device over ethernet with socat for ip $COMFOAIR_IP:$COMFOAIR_PORT"
    /usr/bin/socat -d -d pty,link="$SERIAL_PORT",raw,group-late=dialout,mode=660 tcp:"$COMFOAIR_IP":"$COMFOAIR_PORT" &
    export SERIAL_DEVICE=/dev/comfoair
else
    echo "don't create serial device over ehternet. enable it with SOCAT=True"
fi

# Start the second process
/usr/bin/python3 /opt/hacomfoairmqtt/ca350.py &
```

It seems the second process is dying, because /dev/ttyUSB0 is not accessible as you described earlier. So the container also crashes and is spinning up again a little bit later.

If you have the knowledge, you can try to manipulate the start.sh script and comment out the second process. So you should reach a persistent container start for further debugging.

Afterwards you can execute the command manually

```
/usr/bin/python3 /opt/hacomfoairmqtt/ca350.py
```

Maybe you get some more errors or events to dig deeper. As I have no local attached USB/serial device, I could not provide many support by testing. 

Could you provider more information with "ls -la /dev/ | grep ttyUSB0"? Perhaps there is really a permission problem.

**Reply by Buddeli** [reply]
Here the information about the permission. For the other stuff I do need a little bit more time. :)
```
buddy@dockerversuch:~$ ls -la /dev/ | grep ttyUSB0
crw-rw-rw-  1 root dialout 188,   0 Aug  4 19:50 ttyUSB0
```

**Reply by Buddeli** [reply]
Just for info. i'm a beginner when it comes to programming. but i'll try to follow your steps. 
Regarding your first question about config.ini. Unfortunately I can't find it under this path. Could it be stored somewhere else?

Regarding the SOCAT=true status in the start.sh:
I have commented out the second part and executed the command. I received the following response. 
```
buddy@dockerversuch:~/hacomfoairmqtt/src$ /usr/bin/python3 ca350.py
Traceback (most recent call last):
  File "/home/buddy/hacomfoairmqtt/src/ca350.py", line 22, in <module>
    import paho.mqtt.client as mqtt
ModuleNotFoundError: No module named 'paho'
``` 

I don't think that's what you mean. I have the feeling that I'm changing something in the wrong directory.

**Reply by adorobis** [reply]
You need to execute it in the container. Here it looks like you are running it in the Docker host. So first run: `docker exec -it [yourcontainername] bash`. But if your container constantly crashes you need to update the `start.sh` first so that it does not start the ca350.py script. The script is located in `/usr/local/bin/start.sh` of the container. But I don't know how to do it when the container crashes constantly. Maybe @jschanz could help here.

**Comment by Buddeli** [comment]
After a good night's sleep, I reconsidered the whole thing. Maybe there is a simpler solution for my problem and you don't have to advise a beginner in Docker, VMs and USB configuration. :) 

Here is the initial situation:
My Zehnder ventilation is in the same room as my miniserver (Proxmox with the VMs, Docker etc.) and my Homeassistant which runs on a Raspberry Pi4 is also in the same room. Therefore my idea to connect the ventilation directly via cable. The aim is to control the ventilation via Homeassistant. 

Do you have any simpler solutions such as:
1. connecting the USB to serial adapter directly to the Rasperry (HA)
2. a solution via a microcontroller with ESPHome
3. other simpler ideas?

Maybe that makes it easier?

**Reply by adorobis** [reply]
For option 1 here are the instructions:
https://github.com/adorobis/hacomfoairmqtt/wiki/Home-Assistant-Operating-System
For option 2 there is a recent post using a very nice device: https://community.home-assistant.io/t/zehnder-comfoair-ca350-integration-via-serial-connection-rs232-and-mqtt/173243/434?u=adorobis
Easiest in my opinion will be to fix your current setup as you have everything in place - just something wrong with the usb passing to the container :)

**Reply by Buddeli** [reply]
Thanks for the two options. I tried the first one and realized that something was wrong with my USB adapter. Now I have ordered a new one and voilà, it works. As soon as I have some time, I will try again to connect the adapter via the Docker container.
Thanks for the great script and i will keep you up to date.
